mt_2048.cpp

```diff
@@ -0,0 +1,310 @@
+                            l_state[4 * k + j] = l_state[4 * i + j];
+                            k++;
+                            if (can_move) {
+                                if_move = true;
+                            }
+                        } else {
+                            can_move = true;
+                        }
+                    }
+                    while (k < 4) {
+                        l_state[4 * k + j] = 0;
+                        k++;
+                    }
+                } else
+                    for (i = 0; i < 3; i++) {
+                        if (l_state[4 * i + j] &&
+                            l_state[4 * i + j] == l_state[4 * (i + 1) + j]) {
+                            l_state[4 * i + j] += 1;
+                            score += 1;
+                            l_state[4 * (i + 1) + j] = 0;
+                            i++;
+                        }
+                    }
+            }
+        }
+    }
+    if (!if_move && score == 0) {
+        score = -1;
+    }
+    return score;
+}
+
+int isOver(const int *state) {
+    int zero_num = 0;
+    bool can_combine = false;
+    for (int i = 0; i < 4; i++) {
+        for (int j = 0; j < 4; j++) {
+            if (state[4 * i + j] == 0) zero_num += 1;
+            if ((j < 3 && state[4 * i + j] == state[4 * i + j + 1]) ||
+                (i < 3 && state[4 * i + j] == state[4 * (i + 1) + j])) {
+                can_combine = true;
+            }
+        }
+    }
+    if (zero_num == 0) {
+        if (!can_combine) {
+            return -1;
+        } else {
+            return zero_num;
+        }
+    }
+    return zero_num;
+}
+
+void newBlock(int *state) {
+    int new_num = 1;
+    int r_num = rand() % 10;
+    if (r_num > 8) new_num = 2;
+    bool ok = false;
+    while (!ok) {
+        int r_place = rand() % 16;
+        if (state[r_place] == 0) {
+            state[r_place] = new_num;
+            ok = true;
+        }
+    }
+}
+
+void runGame(const int *d_state, int *d_result, int search_depth, int tid) {
+    int init_state = tid;
+    int depth = search_depth;
+
+    int l_state[16];
+    for (int i = 0; i < 16; i++) l_state[i] = d_state[i];
+    int result;
+    int score = 0;
+    int turn = 0;
+    int zero_num = isOver(l_state);
+    while (depth && zero_num != -1) {
+        turn = init_state % 4;
+        init_state /= 4;
+        depth -= 1;
+        result = combine(l_state, turn);
+        if (result == -1 && depth == search_depth - 1) score -= 100;
+        zero_num = isOver(l_state);
+        score += result * (result > 0);
+        if (zero_num > 0 && result != -1) newBlock(l_state);
+    }
+
+    int count = 0;
+    while (zero_num != -1) {
+        count += 1;
+        turn = rand() % 4;
+        result = combine(l_state, turn);
+        score += result * (result > 0);
+        zero_num = isOver(l_state);
+        if (zero_num > 0 && result != -1) newBlock(l_state);
+    }
+    d_result[tid] = score;
```

Comment:
I think here should be d_result[tid += score. Just like “atotic_add” in my .cu implement

Reply:
You are right! :laughing: